Link unstable features

albertlarsan68 opened the pull request:
Helps with #103548

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3885245337/jobs/6628878190) [(plain)](https://github.com/rust-lang/rust/commit/3e3ea5194b487102d8a9a12b066c427ec0a7dd7e/checks/10556641886/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 050747f69934
Successfully tagged rust-ci:latest
Built container sha256:050747f6993403997e50955f3fdfa895bb534fef75fbc063f41bfe21db96c0de
Uploading finished image to https://ci-caches.rust-lang.org/docker/b8cdc76b9b2adf6aaac1172d56fab8a31470587d5348a0014d9e27d5bcbbe8577c6dad29501e9359634fe46a53fc6079832ca8161c7036f56b2737f782f779b6
upload failed: - to s3://rust-lang-ci-sccache2/docker/b8cdc76b9b2adf6aaac1172d56fab8a31470587d5348a0014d9e27d5bcbbe8577c6dad29501e9359634fe46a53fc6079832ca8161c7036f56b2737f782f779b6 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
tidy: Skipping binary file check, read-only filesystem
Found 505 error codes
Highest error code: `E0791`
* 392 features
thread '<unnamed>' panicked at 'could not read lib feature file /checkout/src/doc/unstable-book/src/library-features/allocator_api.md: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/tidy/src/unstable_book.rs:195:54
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'a scoped thread panicked', src/tools/tidy/src/main.rs:54:5
```

</details>

albertlarsan68 commented:
The only problem that is left, is that the docs of the crates `core`, `alloc`, `std`, `proc_macro`, `test` are re-built every time `x test tidy` is run.

It is just the docs for those crates, not their dependencies, but I think the caching is broken some[how,where]?

bors commented:
:umbrella: The latest upstream changes (presumably #106757) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #107143) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

albertlarsan68 commented:
@jyn514 I rebased and added the missing data to make this change pass.
@rustbot ready

jyn514 commented on src/bootstrap/builder/tests.rs:
```diff
@@ -286,7 +286,10 @@ mod dist {
         // Make sure rustdoc is only built once.
         assert_eq!(
             first(cache.all::<tool::Rustdoc>()),
-            &[tool::Rustdoc { compiler: Compiler { host: a, stage: 2 } },]
+            &[
+                tool::Rustdoc { compiler: Compiler { host: a, stage: 0 } },
+                tool::Rustdoc { compiler: Compiler { host: a, stage: 2 } }
+            ]
```
This doesn't look right. First, the comment is now incorrect, but more importantly, it's important that we only distribute things that have been built with `--stage 2`; std uses `cfg(bootstrap)` so anything built with stage 0 can be just completely different.

I think instead you should change `UnstableBookGen` to use `builder.top_stage` instead of stage 0.

albertlarsan68 replied:
I did this in `UnstableBookGen`, but doing this in Tidy results in stage1 being built, so I keep it stage 0 in tidy.

jyn514 commented on src/doc/unstable-book/src/library-features/allocator-api.md:
```diff
@@ -2,6 +2,12 @@
 
+Items:
+
+- [core::alloc::Allocator](https://doc.rust-lang.org/nightly/core/alloc/trait.Allocator.html)
+- [alloc::alloc::Global](https://doc.rust-lang.org/nightly/alloc/alloc/struct.Global.html)
+- [core::alloc::AllocError](https://doc.rust-lang.org/nightly/core/alloc/struct.AllocError.html)
```
~~Hmm, I thought we'd discussed automatically generating these so they don't get outdated?~~

never mind, I see that tidy suggests these automatically if they get outdated. But then why do some of the files below use `{items}` instead?

albertlarsan68 replied:
It is because each feature does not have an explicit file, most of them are actually autogenerated in the `unstable-book-gen` tool.
I made it so that Tidy errors when an explicitly written feature misses an item and silently ignores the missing features.
In the `unstable-book-gen` tool, I do the same work as Tidy, but this time to include it in the autogenerated page.

jyn514 reviewed (commented):
Can you see how long it takes to run tidy before and after this change? Please include the time it takes to run rustdoc -wjson on the standard library.

jyn514 commented:
Sorry for the long delay btw; if you want to reassign this to Mark I'm ok with that.

jyn514 reviewed (approved):
this looks broadly right to me except for the stage issue :)

albertlarsan68 commented:
```
$ hyperfine -L rev dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2,005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf -s "git checkout {rev};x test tidy;x clean" -w 2 -p "x clean" -m 5 -n "x test tidy of {rev}" "x test tidy"

Benchmark 1: x test tidy of dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2
  Time (mean ± σ):     83.100 s ±  0.655 s    [User: 257.193 s, System: 23.844 s]
  Range (min … max):   82.476 s … 84.087 s    5 runs

Benchmark 2: x test tidy of 005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf
  Time (mean ± σ):     48.405 s ±  0.576 s    [User: 171.333 s, System: 18.779 s]
  Range (min … max):   47.721 s … 49.040 s    5 runs

Summary
  'x test tidy of 005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf' ran
    1.72 ± 0.02 times faster than 'x test tidy of dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2'
```

With dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2 being the tip of this PR, and 005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf being the merge-base between this PR and master.

jyn514 commented:
Oh right, I'd forgotten that you have a slow laptop 😓 

Hmm, given that it's such a large slowdown, maybe we should move the check out of tidy? We can put it in the `test::UnstableBook` step (you'll have to pull it out of the `test_book` macro, but that seems ok).

albertlarsan68 commented:
> Oh right, I'd forgotten that you have a slow laptop 😓

The benchmarks were ran on a dev-desktop.

jyn514 commented:
OH so that includes the time to build bootstrap and tidy in the benchmark. Ok, then 35 additional seconds definitely seems too long.

albertlarsan68 commented:
I should learn to write full messages before posting them, and not have to edit.
I am benchmarking when not cleaning every time.

albertlarsan68 commented:
```
$ hyperfine -L rev dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2,005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf -s "git checkout {rev};x test tidy;x clean" -w 2 -m 5 -n "x test tidy of {rev}" "x test tidy"

Benchmark 1: x test tidy of dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2
  Time (mean ± σ):     14.397 s ±  0.053 s    [User: 30.262 s, System: 3.714 s]
  Range (min … max):   14.328 s … 14.473 s    5 runs

Benchmark 2: x test tidy of 005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf
  Time (mean ± σ):      2.178 s ±  0.033 s    [User: 18.719 s, System: 2.748 s]
  Range (min … max):    2.145 s …  2.219 s    5 runs

Summary
  'x test tidy of 005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf' ran
    6.61 ± 0.10 times faster than 'x test tidy of dfb40c3d0b34f2ad4bc72b4cf8645cd872928be2'
```
The same as above, but not cleaning between each measurement.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4005898015/jobs/6876802144) [(plain)](https://github.com/rust-lang/rust/commit/f4a31073ffee8218b2aad2cf26090826db40045f/checks/10876146824/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:7873766cb4a6a0bb00c54496556b38db3fffa5d6)
Complete job name: PR (x86_64-gnu-llvm-13, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-13
---
   Compiling autocfg v1.1.0
   Compiling num-traits v0.2.12
   Compiling unstable-book-gen v0.1.0 (/checkout/src/tools/unstable-book-gen)
    Finished release [optimized] target(s) in 2.39s
thread 'main' panicked at 'failed to parse JSON docs: Error("missing field `variant_kind`", line: 1, column: 2997)', src/tools/tidy/src/unstable_book.rs:54:59
Build completed unsuccessfully in 0:21:52
```

</details>

albertlarsan68 committed:
Generate unstable docs with items under feature

Add the items to the autogenerated stubs
Make unstable-book-gen use the top stage instead of stage 0
Generate the full docs, only print and save optional notes

albertlarsan68 commented:
@jyn514 I have moved the generation of the pages to `unstable-book-gen`, so now the `src/doc/unstable-book/src/library-features` only contains the optional notes about the feature. So now, there is no tidy changes, and the style is consistent.
I switched to using the in-tree `rustdoc-json-types`, because they are being generated with stage 2 in CI. This means that a simple `x doc unstable-book` will be likely to fail, but I think not having to have logic to handle two versions and bump the `rustdoc-types` dependency at each bootstrap compile bump is a plus. I had this logic at one point, but I removed it to support only the in-tree `rustdoc-json-types`. Do you want me to add it back?
I will try to make a diff of the markdown output of `unstable-book-gen` with and without this change, but the results are likely to be fuzzy / irrelevant.

jyn514 commented:
I don't think I'll have time to review this in the near future, unfortunately.

r? bootstrap

albertlarsan68 commented:
No problem, do what you think is the best.

Welcome @ozkanonur to this PR!

ozkanonur commented on src/tools/unstable-book-gen/src/main.rs:
```diff
@@ -11,13 +13,121 @@ use tidy::unstable_book::{
+                    _ => return None,
+                };
+
+                // Given a `NestedMeta` like `feature = "xyz"`, returns `xyz`.
+                let get_feature_name = |nested: &_| {
+                    match nested {
+                        NestedMeta::Meta(Meta::NameValue(name_value)) => {
+                            if !is_ident(name_value.path.get_ident()?, "feature") {
+                                return None;
+                            }
+                            match &name_value.lit {
+                                Lit::Str(s) => Some(s.value()),
+                                _ => None,
+                            }
+                        }
+                        _ => None,
+                    }
+                };
```
This can be defined outside of the loop.

Also, we can reduce the scopes with the following way
```rs
// Given a `NestedMeta` like `feature = "xyz"`, returns `xyz`.
let get_feature_name = |nested: &_| match nested {
    NestedMeta::Meta(Meta::NameValue(name_value)) => {
        if !is_ident(name_value.path.get_ident()?, "feature") {
            return None;
        }
        match &name_value.lit {
            Lit::Str(s) => Some(s.value()),
            _ => None,
        }
    }
    _ => None,
};
```

ozkanonur commented on src/tools/unstable-book-gen/src/main.rs:
```diff
@@ -11,13 +13,121 @@ use tidy::unstable_book::{
+use syn::{parse::Parser, Ident, Lit, Meta, NestedMeta};
+
+// (item path, doc url)
+fn full_path(krate: &Crate, item: &Id) -> Option<(String, String)> {
+    let item_summary = krate.paths.get(item)?;
+    let kind = &item_summary.kind;
+    let kind_str = serde_json::to_string(kind).ok()?;
+    let mut url = String::from("https://doc.rust-lang.org/nightly/");
+    let mut iter = item_summary.path.iter();
+    iter.next_back();
+    url.push_str(&iter.cloned().collect::<Vec<_>>().join("/"));
+    url.push('/');
+    url.push_str(kind_str.trim_matches('"'));
+    url.push('.');
+    url.push_str(item_summary.path.last().unwrap());
+    url.push_str(".html");
+    Some((item_summary.path.join("::"), url))
+}
```
Is it only me that this looks hard to debug/understand the workflow inside? :)

Maybe we can have some comments to inform developer about what's going on in there. + Using `expect` instead of `unwrap` can be nice to have as well.

albertlarsan68 replied:
This tries (quite badly) to generate the URL that the item belongs to.
If there is a way to get it directly from Rustdoc, I will happily use it.
Concerning the unwraps, this is mostly prototype code.

ozkanonur replied:
> This tries (quite badly) to generate the URL that the item belongs to.
If there is a way to get it directly from Rustdoc, I will happily use it.

Having comments which explains aim of the function instructions(which are complicated to read/track) can be handy imo.(e.g https://github.com/rust-lang/rust/blob/91eb6f9acfcfde6832d547959ad2d95b1ac0b5dc/src/bootstrap/builder.rs#L1165-L1930)

> Concerning the unwraps, this is mostly prototyping code that is not meant to last.

It's just to give more clear information in context. Not so necessary I guess.

onur-ozkan reviewed (changes requested):
Greatk work! :fire:

I have some suggestions for this implementation:

albertlarsan68 committed:
Make the link generation work for more cases

bors commented:
:umbrella: The latest upstream changes (presumably #113508) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ozkanonur commented:
@albertlarsan68 do we have any plans for this PR?

albertlarsan68 commented:
The thing that I am having a hard time with is generating the links. The machinery to detect what is gated is in place, but the links are hard to generate.

Dylan-DPC commented:
@albertlarsan68 any updates on it?

albertlarsan68 commented:
Nothing new.
This PR has the code needed to generate at least the (human-readable) paths to unstable items, but I wasn't able to generate correct links to them.
Maybe I'll remove the link generation code, then we could merge this as a "helps with" instead of a "fixes".

albertlarsan68 committed:
Finish rebasing

bors commented:
:umbrella: The latest upstream changes (presumably #120491) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@albertlarsan68 if you can rebase this and resolve the conflicts , we can push this forward for a review

JohnCSimon commented:
@albertlarsan68 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
